File: 2024/day_04/part_2.lua
-- https://adventofcode.com/2024/day/4

local function read_input_file(filename)
  local dir = debug.getinfo(1, 'S').source:sub(2):match('^(.*)/')
  local file, err
  if dir == nil then
    file, err = io.open(filename)
  else
    file, err = io.open(dir .. '/' .. filename)
  end
  if not file then
    print('input opening failed ' .. err)
    os.exit()
  end
  return file
end

local function primary_diagonal_mas(grid, i, j)
  if
    (i - 1) >= 1
    and (j - 1) >= 1
    and grid[i - 1]:sub(j - 1, j - 1) == 'M'
    and (i + 1) <= #grid
    and (j + 1) <= grid[i]:len()
    and grid[i + 1]:sub(j + 1, j + 1) == 'S'
  then
    return true
  end
  return false
end

local function primary_diagonal_sam(grid, i, j)
  if
    (i - 1) >= 1
    and (j - 1) >= 1
    and grid[i - 1]:sub(j - 1, j - 1) == 'S'
    and (i + 1) <= #grid
    and (j + 1) <= grid[i]:len()
    and grid[i + 1]:sub(j + 1, j + 1) == 'M'
  then
    return true
  end
  return false
end

local function secondary_diagonal_mas(grid, i, j)
  if
    (i - 1) >= 1
    and (j + 1) <= grid[i]:len()
    and grid[i - 1]:sub(j + 1, j + 1) == 'M'
    and (i + 1) <= #grid
    and (j - 1) >= 1
    and grid[i + 1]:sub(j - 1, j - 1) == 'S'
  then
    return true
  end
  return false
end

local function secondary_diagonal_sam(grid, i, j)
  if
    (i - 1) >= 1
    and (j + 1) <= grid[i]:len()
    and grid[i - 1]:sub(j + 1, j + 1) == 'S'
    and (i + 1) <= #grid
    and (j - 1) >= 1
    and grid[i + 1]:sub(j - 1, j - 1) == 'M'
  then
    return true
  end
  return false
end

local function main()
  local result = 0
  local grid = {}
  local file = read_input_file('input.txt')
  for line in file:lines() do
    table.insert(grid, line)
  end
  file:close()

  for i = 1, #grid do
    for j = 1, grid[i]:len() do
      if grid[i]:sub(j, j) == 'A' then
        if
          (primary_diagonal_mas(grid, i, j) or primary_diagonal_sam(grid, i, j))
          and (secondary_diagonal_sam(grid, i, j) or secondary_diagonal_mas(grid, i, j))
        then
          result = result + 1
        end
      end
    end
  end

  print(result)
end

main()

File: 2024/day_10/part_2.py
""" https://adventofcode.com/2024/day/10 """

import os


def get_file_path(filename: str) -> str:
    dir_path = os.path.split(os.path.abspath(__file__))[0]
    return os.path.join(dir_path, filename)


def search(row: int, col: int, grid: list[list[int]]) -> int:
    if grid[row][col] == 9:
        return 1
    score = 0
    for dr, dc in [(-1, 0), (1, 0), (0, 1), (0, -1)]:
        nr, nc = row + dr, col + dc
        if nr < 0 or nc < 0 or nr >= len(grid) or nc >= len(grid[nr]):
            continue
        if grid[nr][nc] - grid[row][col] == 1:
            score += search(nr, nc, grid)
    return score


def main():
    grid: list[list[int]]

    with open(get_file_path("input.txt")) as fd:
        grid = [list(map(int, list(line.rstrip()))) for line in fd.readlines()]

    result = 0
    for i in range(len(grid)):
        for j in range(len(grid[i])):
            if grid[i][j] == 0:
                result += search(i, j, grid)

    print(result)


main()
